allennlp/models/semantic_parsing/atis/atis_semantic_parser.py

```diff
@@ -0,0 +1,592 @@
+            # productions of that non-terminal.  We'll first split those productions by global vs.
+            # linked action.
+
+            action_indices = [action_map[action_string] for action_string in action_strings]
+            production_rule_arrays = [(possible_actions[index], index) for index in action_indices]
+            global_actions = []
+            linked_actions = []
+            for production_rule_array, action_index in production_rule_arrays:
+                if production_rule_array[1]:
+                    global_actions.append((production_rule_array[2], action_index))
+                else:
+                    linked_actions.append((production_rule_array[0], action_index))
+
+            if global_actions:
+                global_action_tensors, global_action_ids = zip(*global_actions)
+                global_action_tensor = entity_types.new_tensor(torch.cat(global_action_tensors, dim=0),
+                                                               dtype=torch.long)
+                global_input_embeddings = self._action_embedder(global_action_tensor)
+                global_output_embeddings = self._output_action_embedder(global_action_tensor)
+                translated_valid_actions[key]['global'] = (global_input_embeddings,
+                                                           global_output_embeddings,
+                                                           list(global_action_ids))
+            if linked_actions:
+                linked_rules, linked_action_ids = zip(*linked_actions)
+                entities = linked_rules
+                entity_ids = [entity_map[entity] for entity in entities]
+                entity_linking_scores = linking_scores[entity_ids]
+                entity_type_tensor = entity_types[entity_ids]
+                entity_type_embeddings = self._entity_type_decoder_embedding(entity_type_tensor)
+                entity_type_embeddings = entity_types.new_tensor(entity_type_embeddings, dtype=torch.float)
+                translated_valid_actions[key]['linked'] = (entity_linking_scores,
+                                                           entity_type_embeddings,
+                                                           list(linked_action_ids))
+
+        return GrammarStatelet(['statement'],
+                               {},
+                               translated_valid_actions,
+                               {},
+                               self.is_nonterminal,
+                               reverse_productions=False)
+
+    @overrides
+    def decode(self, output_dict: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
+        """
+        This method overrides ``Model.decode``, which gets called after ``Model.forward``, at test
+        time, to finalize predictions.  This is (confusingly) a separate notion from the "decoder"
+        in "encoder/decoder", where that decoder logic lives in ``GrammarBasedState``.
```

Comment:
Hmm, the decoder logic lives in the `TransitionFunction`.  If we say this in other places, we should fix it there too.